core/spec/models/spree/promotion_code/code_builder_spec.rb

```diff
@@ -0,0 +1,67 @@
+require 'spec_helper'
+
+describe Spree::PromotionCode::CodeBuilder do
+  let(:promotion) { build_stubbed :promotion }
+  let(:base_code) { "abc" }
+  let(:builder) do
+    described_class.new promotion, base_code, num_codes
+  end
+
+  describe "#build_promotion_codes" do
+    subject { builder.build_promotion_codes }
+
+    context "with one code" do
+      let(:num_codes) { 1 }
+
+      it "builds a single promotion code" do
+        subject
+        expect(builder.promotion.codes).to have(num_codes).items
+      end
+
+      it "creates the promotion code with the correct value" do
+        subject
+        expect(builder.promotion.codes.first.value).to eq base_code
+      end
+    end
+
+    context "with more than one code" do
+      before { srand 123 }
+      let(:num_codes) { 2 }
+
+      it "builds the correct number of codes" do
+        subject
+        expect(builder.promotion.codes).to have(num_codes).items
+      end
+
+      it "builds codes with distinct values" do
+        subject
+        expect(builder.promotion.codes.map(&:value).uniq).to have(num_codes).items
+      end
+
+      it "builds codes with the same base prefix" do
+        subject
+        values = builder.promotion.codes.map &:value
+        expect(values.all? { |val| val.starts_with?("#{base_code}_") }).to be true
+      end
+
+      context 'when there is a code colision' do
+        before do
+          @old_length = described_class.default_random_code_length
+          described_class.default_random_code_length = 1
+        end
+        after { described_class.default_random_code_length = @old_length }
+
+        let(:number_of_codes) { 5 }
```

Comment:
I'd still be interested in making this 26 and getting it to pass. I know it may not be 100% useful for everyone to be able to use all the space in the unique string list for a particular code-length, but it would be nice to know that whatever implementation we decide on can handle collisions up to the very end.